[Bug #77120] Handle OCI_SUCCESS_WITH_INFO returned from OCISessionBegin

morozov opened the pull request:
Create a test user:
```sql
CREATE PROFILE test_profile LIMIT PASSWORD_LIFE_TIME 1/86400 PASSWORD_GRACE_TIME 7;
CREATE USER bob IDENTIFIED BY password PROFILE test_profile;
GRANT CREATE SESSION TO bob;
```

Run the script:
```php
$conn = new PDO('oci:dbname=(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(HOST=localhost)(PORT=1521))(CONNECT_DATA=(SID=XE)))', 'bob', 'password');
```

**Expected result**:
The connection is successfully established. The details of the informational message are available via `PDO::errorInfo()`.

**Actual result**:
```
PDOException: SQLSTATE[HY000]: OCISessionBegin: OCI_SUCCESS_WITH_INFO: ORA-28002: the password will expire within 7 days
 (ext/pdo_oci/oci_driver.c:685) in test.php on line 3
```

**Change summary**:
1. Proceed with connection initialization if `OCI_SUCCESS_WITH_INFO` is returned.
2. Do not throw an exception upon `OCI_SUCCESS_WITH_INFO`.

nikic commented:
cc @cjbj

morozov commented:
@nikic is there a way to proceed given the absence of @cjbj's feedback?

cjbj commented:
I'd like to review it first.

ramsey commented:
@morozov Is it possible to provide a test with this?

morozov commented:
Absolutely. I just want to make sure the fix as such is acceptable.

cjbj commented:
Adding a test now would expedite me reviewing it.  Otherwise I have to create the test... :)

morozov commented:
@cjbj, @ramsey please see the test.

cjbj commented:
@morozov Thank you.  I should get time later this week to take a look.

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,56 @@
+
+$conn->exec(<<<'SQL'
+BEGIN
+    EXECUTE IMMEDIATE 'DROP PROFILE TEST_PROFILE CASCADE';
```
Best use names that are more unique, like bug77120_profile

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,56 @@
+
+$conn->exec(<<<'SQL'
+BEGIN
+    EXECUTE IMMEDIATE 'DROP USER BOB';
```
Can you use a name that is more unique? And add CASCADE?

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,56 @@
+
+$conn = new PDO(getenv('PDOTEST_DSN'), 'BOB', 'Passw0rd');
+var_dump($conn->errorInfo());
+?>
```
The newly created user needs to be dropped at the end of the test.

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,56 @@
+  [1]=>
+  int(28002)
+  [2]=>
+  string(%d) "OCISessionBegin: OCI_SUCCESS_WITH_INFO: ORA-28002: the password will expire within %d days
```
Use a %s for the English text so that the test passes when messages are not in English.

morozov replied:
Done. But I'd say it makes more sense to enforce the English locale when connecting to the DB.

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,56 @@
+);
+
+$conn->exec('CREATE PROFILE TEST_PROFILE LIMIT PASSWORD_LIFE_TIME 1/86400 PASSWORD_GRACE_TIME 1');
+$conn->exec('CREATE USER BOB IDENTIFIED BY Passw0rd PROFILE TEST_PROFILE');
```
It's best to create a random password, even for tests.

cjbj reviewed (changes requested):
Can you also check some other SUCCESS_WITH_INFO cases such as PL/SQL compilation failures?  Check them for various PDO::ATTR_ERRMODE values. (And verify the old & new behaviors).  Thank you!

morozov commented:
> Can you also check some other SUCCESS_WITH_INFO cases such as PL/SQL compilation failures? Check them for various PDO::ATTR_ERRMODE values.

What would be the best way to reproduce this? I tried a couple of examples I found online (e.g. [this](http://www.dba-oracle.com/t_ora_24344_success_with_compilation_error.htm)) in PhpStorm but it didn't trigger the expected error.

morozov commented:
I manages to reproduce the compilation error with the following:
```php
function triggerCompilationError(PDO $conn): void {
    $conn->exec(<<<'SQL'
CREATE OR REPLACE FUNCTION BUG77120(INT A) RETURN INT
AS
BEGIN
    RETURN 0;
END;
SQL
    );
}

triggerCompilationError($conn);

$conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
triggerCompilationError($conn);
```
Which produces the following:

> Warning: PDO::exec(): SQLSTATE[HY000]: General error: 24344 OCIStmtExecute: OCI_SUCCESS_WITH_INFO: ORA-24344: success with compilation error
 (ext/pdo_oci/oci_driver.c:340) in ext/pdo_oci/tests/oci_success_with_info.php on line 41
> 
> Fatal error: Uncaught PDOException: SQLSTATE[HY000]: General error: 24344 OCIStmtExecute: OCI_SUCCESS_WITH_INFO: ORA-24344: success with compilation error
 (ext/pdo_oci/oci_driver.c:340) in ext/pdo_oci/tests/oci_success_with_info.php:41

I assume the error handling still exists in this scenario but it's out of the scope of the original issue.

cjbj commented:
> I assume the error handling still exists in this scenario but it's out of the scope of the original issue. Statement error handling can be configured at runtime while connection error handling cannot.

The PR changes `_oci_error()` which is called in oci_statement.c so it seems wiser to check that any existing statement error handling isn't borked.

You could use `create or replace procedure p() as begin null; end;` to create a compilation error.

morozov commented:
> The PR changes `_oci_error()` which is called in oci_statement.c so it seems wiser to check that any existing statement error handling isn't borked.

It wasn't borked, it worked as previously, see https://github.com/php/php-src/pull/6857#issuecomment-842656999.

Please see the update. I modified handling `OCI_SUCCESS_WITH_INFO` returned from `OCIStmtExecute` and updated the test.

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,116 @@
+  [1]=>
+  int(24344)
+  [2]=>
+  string(143) "OCIStmtExecute: OCI_SUCCESS_WITH_INFO: ORA-24344: %s
```
This failed for me until I changed 143 to %d

cjbj commented on ext/pdo_oci/tests/oci_success_with_info.phpt:
```diff
@@ -0,0 +1,116 @@
+  [1]=>
+  int(24344)
+  [2]=>
+  string(143) "OCIStmtExecute: OCI_SUCCESS_WITH_INFO: ORA-24344: %s
```
This failed for me until I changed 143 to %d

cjbj reviewed (changes requested):
Thanks for tidying up, and for adding the extra checks to confirm the behavior.
With the trivial test changes I noted, let's get this merged.
Thanks for making PDO_OCI better.

morozov commented:
@cjbj any chance you could review this again?

cjbj commented:
LGTM

@ramsey can you advise on the (new) current process for merging?

morozov commented:
@ramsey what would be the next step to get this patch accepted?

ramsey commented:
All good. Sorry for the delay. Thanks!

cmb69 commented:
This should have an entry in NEWS.

cjbj commented:
> This should have an entry in NEWS.

Done.  Thanks all.

ramsey commented:
Thanks, @cjbj!
